% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={Test DoseRider},
  pdfauthor={Pablo Monfort},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{Test DoseRider}
\author{Pablo Monfort}
\date{2023-06-29}

\begin{document}
\maketitle

\begin{verbatim}
## Loading required package: nlme
\end{verbatim}

\begin{verbatim}
## This is mgcv 1.8-42. For overview type 'help("mgcv-package")'.
\end{verbatim}

\begin{verbatim}
## Loading required package: MatrixGenerics
\end{verbatim}

\begin{verbatim}
## Loading required package: matrixStats
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'MatrixGenerics'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:matrixStats':
## 
##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
##     colWeightedMeans, colWeightedMedians, colWeightedSds,
##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
##     rowWeightedSds, rowWeightedVars
\end{verbatim}

\begin{verbatim}
## Loading required package: GenomicRanges
\end{verbatim}

\begin{verbatim}
## Loading required package: stats4
\end{verbatim}

\begin{verbatim}
## Loading required package: BiocGenerics
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'BiocGenerics'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,
##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which.max, which.min
\end{verbatim}

\begin{verbatim}
## Loading required package: S4Vectors
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'S4Vectors'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     expand.grid, I, unname
\end{verbatim}

\begin{verbatim}
## Loading required package: IRanges
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'IRanges'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:nlme':
## 
##     collapse
\end{verbatim}

\begin{verbatim}
## Loading required package: GenomeInfoDb
\end{verbatim}

\begin{verbatim}
## Loading required package: Biobase
\end{verbatim}

\begin{verbatim}
## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'Biobase'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:MatrixGenerics':
## 
##     rowMedians
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:matrixStats':
## 
##     anyMissing, rowMedians
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'dplyr'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:Biobase':
## 
##     combine
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:GenomicRanges':
## 
##     intersect, setdiff, union
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:GenomeInfoDb':
## 
##     intersect
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:IRanges':
## 
##     collapse, desc, intersect, setdiff, slice, union
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:S4Vectors':
## 
##     first, intersect, rename, setdiff, setequal, union
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:BiocGenerics':
## 
##     combine, intersect, setdiff, union
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:matrixStats':
## 
##     count
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:nlme':
## 
##     collapse
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:stats':
## 
##     filter, lag
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Create a GAMM formula:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Define the formulas for the models}

\CommentTok{\# Define the formulas for the models}
\NormalTok{base\_formula }\OtherTok{\textless{}{-}} \FunctionTok{create\_gamm\_formula}\NormalTok{(}\AttributeTok{response =} \StringTok{"counts"}\NormalTok{,}
                                    \AttributeTok{fixed\_effects =} \StringTok{"dose"}\NormalTok{,}
                                    \AttributeTok{random\_effects =} \StringTok{"gene"}\NormalTok{,}
                                    \AttributeTok{model\_type =} \StringTok{"base"}\NormalTok{)}

\NormalTok{linear\_formula }\OtherTok{\textless{}{-}} \FunctionTok{create\_gamm\_formula}\NormalTok{(}\AttributeTok{response =} \StringTok{"counts"}\NormalTok{,}
                                      \AttributeTok{fixed\_effects =} \StringTok{"dose"}\NormalTok{,}
                                      \AttributeTok{random\_effects =} \StringTok{"gene"}\NormalTok{,}
                                      \AttributeTok{model\_type =} \StringTok{"linear"}\NormalTok{)}

\NormalTok{cubic\_formula }\OtherTok{\textless{}{-}} \FunctionTok{create\_gamm\_formula}\NormalTok{(}\AttributeTok{response =} \StringTok{"counts"}\NormalTok{,}
                                     \AttributeTok{fixed\_effects =} \StringTok{"dose"}\NormalTok{,}
                                     \AttributeTok{random\_effects =} \StringTok{"gene"}\NormalTok{,}
                                     \AttributeTok{model\_type =} \StringTok{"cubic"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  Create a summarized experiment:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{omic }\OtherTok{\textless{}{-}} \StringTok{"proteomic"}
\NormalTok{proteomics\_data }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(}\StringTok{"\textasciitilde{}/Projects/TOX/project\_tox/data/proteomic/BCI\_vsn\_impute\_batch.csv"}\NormalTok{, }\AttributeTok{check.names =}\NormalTok{ F, }\AttributeTok{row.names =} \DecValTok{1}\NormalTok{) }
\NormalTok{metadata }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(}\StringTok{"\textasciitilde{}/Projects/TOX/project\_tox/data/proteomic/target.txt"}\NormalTok{, }\AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{) }
\FunctionTok{colnames}\NormalTok{(metadata) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"SAMPLE"}\NormalTok{, }\StringTok{"CHANNEL"}\NormalTok{, }\StringTok{"EXPERIMENT"}\NormalTok{, }\StringTok{"CELLTYPE"}\NormalTok{, }\StringTok{"CONCENTRATION"}\NormalTok{, }\StringTok{"INDEX"}\NormalTok{,}\StringTok{"GROUP"}\NormalTok{,}\StringTok{"name"}\NormalTok{,}\StringTok{"dose"}\NormalTok{,}\StringTok{"sample"}\NormalTok{)}
\NormalTok{metadata }\OtherTok{\textless{}{-}}\NormalTok{ metadata[metadata}\SpecialCharTok{$}\NormalTok{CELLTYPE }\SpecialCharTok{==} \StringTok{"B"}\NormalTok{,]}
\FunctionTok{rownames}\NormalTok{(metadata) }\OtherTok{\textless{}{-}}\NormalTok{ metadata}\SpecialCharTok{$}\NormalTok{sample}
\NormalTok{proteomics\_data }\OtherTok{\textless{}{-}}\NormalTok{ proteomics\_data[}\FunctionTok{rownames}\NormalTok{(metadata)]}

\NormalTok{se }\OtherTok{\textless{}{-}} \FunctionTok{create\_summarized\_experiment}\NormalTok{(proteomics\_data, metadata)}
\end{Highlighting}
\end{Shaded}

Replace \texttt{proteomics\_data} with your actual proteomics data
object and \texttt{metadata} with the corresponding metadata object.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  Estimate model parameters, only for RNASeq data:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{if}\NormalTok{ ( omic }\SpecialCharTok{==} \StringTok{"rnaseq"}\NormalTok{)\{}
\NormalTok{parameters }\OtherTok{\textless{}{-}} \FunctionTok{estimate\_model\_parameters}\NormalTok{(se, formula)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{3}
\tightlist
\item
  Load gene sets from ConsensusPathDB:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{file\_path }\OtherTok{\textless{}{-}} \StringTok{"../data/CPDB\_pathways\_genes.tab"}
\NormalTok{gmt }\OtherTok{\textless{}{-}} \FunctionTok{load\_consensupathdb\_genesets}\NormalTok{(file\_path)}
\NormalTok{gmt }\OtherTok{\textless{}{-}} \FunctionTok{filter\_gmt\_by\_size}\NormalTok{(}\AttributeTok{gmt =}\NormalTok{ gmt, }\AttributeTok{minGenesetSize =} \DecValTok{200}\NormalTok{, }\AttributeTok{maxGenesetSize =} \DecValTok{1200}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{4}
\tightlist
\item
  Perform the analysis on gene sets:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{res }\OtherTok{\textless{}{-}} \FunctionTok{perform\_analysis}\NormalTok{(se, gmt, base\_formula, linear\_formula, cubic\_formula, }\AttributeTok{dose\_col =} \StringTok{"dose"}\NormalTok{, }\AttributeTok{sample\_col =} \StringTok{"sample"}\NormalTok{, }\AttributeTok{omic =} \StringTok{"proteomic"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\textbf{Check the results}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{table}\NormalTok{(res}\SpecialCharTok{$}\NormalTok{FDR }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## FALSE  TRUE 
##   100    22
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(res[res}\SpecialCharTok{$}\NormalTok{FDR }\SpecialCharTok{\textless{}} \FloatTok{0.05} \SpecialCharTok{\&} \SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(res}\SpecialCharTok{$}\NormalTok{FDR),])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                                                                     Geneset
## 1                                      Prion disease - Homo sapiens (human)
## 2  Pathways of neurodegeneration - multiple diseases - Homo sapiens (human)
## 4                                     Focal adhesion - Homo sapiens (human)
## 7                            Diabetic cardiomyopathy - Homo sapiens (human)
## 9                                  Alzheimer disease - Homo sapiens (human)
## 11                                Huntington disease - Homo sapiens (human)
##    Geneset_Size Genes  Base_AIC   Base_BIC Linear_AIC Linear_BIC Cubic_AIC
## 1           273   139 -3161.773 -2306.4459  -3175.166 -2313.7242 -3175.162
## 2           475   202 -5314.166 -3998.0166  -5327.849 -4005.2112 -5327.845
## 4           201    72 -1345.259  -947.3144  -1351.716  -948.3159 -1360.207
## 7           203   104 -2684.337 -2073.2024  -2713.480 -2096.5204 -2713.479
## 9           369   165 -4038.758 -2996.0800  -4055.265 -3006.3012 -4055.261
## 11          306   160 -3915.987 -2909.6652  -3935.155 -2922.5775 -3935.155
##    Cubic_BIC P_Value_Linear P_Value_Cubic          FDR
## 1  -2313.706      1.226e-04     2.000e-03 1.284211e-02
## 2  -4005.192      1.058e-04     2.000e-03 1.284211e-02
## 4   -947.135      4.000e-03     2.000e-03 1.284211e-02
## 7  -2096.516      4.678e-08     1.487e-06 2.015711e-05
## 9  -3006.276      2.537e-05     4.987e-04 4.056093e-03
## 11 -2922.575      6.668e-06     1.481e-04 1.379471e-03
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{5}
\tightlist
\item
  Plot smooth curves:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lP }\OtherTok{\textless{}{-}}\FunctionTok{list}\NormalTok{()}
\NormalTok{j }\OtherTok{=} \DecValTok{1}
\ControlFlowTok{for}\NormalTok{ (genest\_name }\ControlFlowTok{in} \FunctionTok{unique}\NormalTok{(res[res}\SpecialCharTok{$}\NormalTok{FDR }\SpecialCharTok{\textless{}} \FloatTok{0.05} \SpecialCharTok{\&} \SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(res}\SpecialCharTok{$}\NormalTok{FDR),]}\SpecialCharTok{$}\NormalTok{Geneset)) \{}
  

\NormalTok{  i }\OtherTok{\textless{}{-}} \FunctionTok{find\_geneset\_index}\NormalTok{(}\AttributeTok{gmt =}\NormalTok{ gmt, }\AttributeTok{geneset\_name =}\NormalTok{ geneset\_name)}
  
\NormalTok{  geneset }\OtherTok{\textless{}{-}}\NormalTok{ gmt[[i]]}\SpecialCharTok{$}\NormalTok{genes}
  
\NormalTok{  long\_df }\OtherTok{\textless{}{-}} \FunctionTok{prepare\_data}\NormalTok{(se, geneset, }\AttributeTok{dose\_col=}\StringTok{"dose"}\NormalTok{, }\AttributeTok{sample\_col =} \StringTok{"sample"}\NormalTok{, }\AttributeTok{omic =} \StringTok{"proteomics"}\NormalTok{)}
  
\NormalTok{  cubic\_results }\OtherTok{\textless{}{-}} \FunctionTok{fit\_gam}\NormalTok{(cubic\_formula, long\_df)}
  
\NormalTok{  p }\OtherTok{\textless{}{-}} \FunctionTok{plot\_smooth}\NormalTok{(cubic\_results, long\_df) }\SpecialCharTok{+} \FunctionTok{ggtitle}\NormalTok{(genest\_name)}
\NormalTok{  lP[[j]] }\OtherTok{\textless{}{-}}\NormalTok{ p}
\NormalTok{  j }\OtherTok{=}\NormalTok{ j }\SpecialCharTok{+} \DecValTok{1}
\NormalTok{\}}
\NormalTok{combined\_plot }\OtherTok{\textless{}{-}}\NormalTok{ cowplot}\SpecialCharTok{::}\FunctionTok{plot\_grid}\NormalTok{(}\AttributeTok{plotlist =}\NormalTok{ lP, }\AttributeTok{ncol =} \DecValTok{3}\NormalTok{)}

\CommentTok{\# Save the combined plot}
\FunctionTok{ggsave}\NormalTok{(}\StringTok{"../../plots/Significant\_Geneset.png"}\NormalTok{, }\AttributeTok{plot =}\NormalTok{ combined\_plot, }\AttributeTok{width =} \DecValTok{22}\NormalTok{, }\AttributeTok{height =} \DecValTok{22}\NormalTok{, }\AttributeTok{limitsize =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{../../plots/Significant_Geneset.png}
\caption{Smooth Curves for Significant Gene Sets. This figure showcases
the smooth curves obtained for the gene sets identified as significant
in the proteomic data analysis. Each subplot represents a distinct gene
set, and the curves portray the relationship between dose and normalized
expression levels. The analysis utilized the cubic model formula to fit
the gene set data and employed a stringent false discovery rate (FDR)
threshold of 0.05 to determine significance.}
\end{figure}

\end{document}
