---
title: "Test DoseRider"
author: "Pablo Monfort"
date: "2023-06-29"
output:
  pdf_document: default
  html_document: default
---
```{r, warning=FALSE, echo=FALSE, include=FALSE}
library(mgcv)
library(SummarizedExperiment)
library(dplyr)
library(ggplot2)
source("doseRider_functions.R")
```



1. Create a GAMM formula:

```{r}
# Define the formulas for the models

# Define the formulas for the models
base_formula <- create_gamm_formula(response = "counts",
                                    fixed_effects = "dose",
                                    random_effects = "gene",
                                    model_type = "base")

linear_formula <- create_gamm_formula(response = "counts",
                                      fixed_effects = "dose",
                                      random_effects = "gene",
                                      model_type = "linear")

cubic_formula <- create_gamm_formula(response = "counts",
                                     fixed_effects = "dose",
                                     random_effects = "gene",
                                     model_type = "cubic")

print(base_formula)
print(linear_formula)
print(cubic_formula)

```

2. Create a summarized experiment:

```{r}
omic <- "proteomic"
proteomics_data <- read.csv(
  "../..//Projects/TOX/project_tox/data/proteomic/BCI_vsn_impute_batch.csv", 
  check.names = F, row.names = 1) 

metadata <- read.csv(
  "../..//Projects/TOX/project_tox/data/proteomic/target.txt", sep = "\t") 

colnames(metadata) <- c("SAMPLE", "CHANNEL", "EXPERIMENT", 
                        "CELLTYPE", "CONCENTRATION", "INDEX","GROUP","name","dose","sample")

metadata <- metadata[metadata$CELLTYPE == "B",]
rownames(metadata) <- metadata$sample
proteomics_data <- proteomics_data[rownames(metadata)]

se <- create_summarized_experiment(proteomics_data, metadata)
```

Replace `proteomics_data` with your actual proteomics data object and `metadata` with the corresponding metadata object.

3. Estimate model parameters, only for RNASeq data:
```{r}
if ( omic == "rnaseq"){
parameters <- estimate_model_parameters(se, formula)
}
```

4. Load gene sets from ConsensusPathDB:

```{r}
file_path <- "../data/CPDB_pathways_genes.tab"
gmt <- load_consensupathdb_genesets(file_path)
gmt <- filter_gmt_by_size(gmt = gmt, minGenesetSize = 200, maxGenesetSize = 1200)
```


5. Perform the analysis on gene sets:
```{r, warning=FALSE, eval = F}
res <- perform_analysis(se, 
                        gmt, 
                        base_formula, 
                        linear_formula, 
                        cubic_formula, 
                        dose_col = "dose", 
                        sample_col = "sample", 
                        omic = "proteomic")

save(res,file = "../data/res.rda")
```

**Check the results**



```{r}
load("../data/res.rda")
table(res$FDR < 0.05)
```

```{r}
head(res[res$FDR < 0.05 & !is.na(res$FDR),], 25)
```

6. Plot smooth curves:

```{r, warning=FALSE, eval=FALSE}
lP <-list()
j = 1
for (geneset_name in unique(res[res$FDR < 0.05 & !is.na(res$FDR),]$Geneset)) {
  
  print(geneset_name)
  i <- find_geneset_index(gmt = gmt, geneset_name = geneset_name)
  
  geneset <- gmt[[i]]$genes
  #print(geneset)
  long_df <- prepare_data(se, geneset, dose_col="dose", 
                          sample_col = "sample", omic = "proteomics")
  
  cubic_results <- fit_gam(cubic_formula, long_df)
  
  p <- plot_smooth(cubic_results, long_df) + ggtitle(geneset_name)
  lP[[j]] <- p
  j = j + 1
}

combined_plot <- cowplot::plot_grid(plotlist = lP, ncol = 5)

# Save the combined plot
ggsave("../../plots/Significant_Geneset.png", plot = combined_plot, width = 22, height = 22, limitsize = FALSE)
```

![Smooth Curves for Significant Gene Sets. This figure showcases the smooth curves obtained for the gene sets identified as significant in the proteomic data analysis. Each subplot represents a distinct gene set, and the curves portray the relationship between dose and normalized expression levels. The analysis utilized the cubic model formula to fit the gene set data and employed a stringent false discovery rate (FDR) threshold of 0.05 to determine significance.](../../plots/Significant_Geneset.png)

7. Obtain Trend Chande Dose (TCD)

