---
title: "Using doseRider for Studying Non-Linear Dose-Response in BPA"
author: "Pablo Monfort-Lanzas"
date: "2023-08-01"
output: html_document
---

# Using doseRider for Studying Non-Linear Dose-Response in BPA

The `doseRider` package is used for analyzing dose-response relationships in gene expression data using Generalized Additive Mixed Models (GAMMs). This report focuses on bisphenol A (BPA) and its alternatives, analyzing their impact on gene expression in MCF-7 cells.

## RMarkdown Workflow for doseRider Analysis

### 1. Loading Required Packages

```{r, warning=FALSE, message=FALSE}
# Load necessary libraries
library(doseRider)
library(DESeq2)
library(ggplot2)
library(dplyr)
# Function to load data
load_data <- function(expr_data_path, metadata_path) {
  expr_data <- read.csv(expr_data_path, sep = "\t", header = TRUE, check.names = FALSE, fill = TRUE, row.names = 1, na.strings = "#NA")
  metadata <- read.csv(metadata_path, sep = "\t", header = TRUE, check.names = FALSE, fill = TRUE)
  list(expr_data = expr_data, metadata = metadata)
}

# Function to create SummarizedExperiment
create_se <- function(data, metadata, omic, dose_col) {
  metadata$sample <- rownames(metadata)
  se <- SummarizedExperiment(assays = as.matrix(data), colData = metadata)
  
  if (omic == "rnaseq") {
    se <- estimate_model_parameters(se)  # Skip DEA but estimate model parameters
  } 
  
  return(se)
}
```

### 2. Data Import

Data from PRJNA869442 (GSE211183) has been preprocessed for this analysis. For this example we only use the BPAF data

```{r}
# Load preprocessed data
data("bpaf_data")
# bpaf_data <- estimate_model_parameters(bpaf_data)
```

Or loading data from the diles

```{r}
metadata_path <- system.file("extdata", "PRJNA869442_BPAF_metadata.tsv", package = "doseRider")
expresion_path <- system.file("extdata", "PRJNA869442_BPAF_expression_data.tsv", package = "doseRider")

# Load data
data_list <- load_data(expresion_path, metadata_path)
# Create SummarizedExperiment
se <- create_se(data = data_list$expr_data, 
                metadata = data_list$metadata,
                omic = "rnaseq",
                dose_col = "Dose")

```


### 3. Load Gene Sets

Several GMT files are avalible int he DoseRider wbesite (https://doserider.i-med.ac.at/). But in this example we will use the High-Response-Toxicogenomics. This custom gene set focuses on genes identified from the Consensus Path Database (CPDB) that exhibit significant changes across all compounds present in the TG-GATES database. This curated gene set emphasizes the most responsive genes involved in toxicological responses, facilitating a targeted approach in dose-response analysis. 

```{r}
# Specify the path to your GMT files
gmt_path <- system.file("extdata", "High-Response-Toxicogenomics.gmt", package = "doseRider")
#gmt_path <- "/home/monfortl/gmt/Homo sapiens/MSigD_C2_CGP_genomics_symbols.gmt"
gmt <- read_gmt(gmt_path)

```

### 4. Dose-Response Model with doseRider

In this example, we utilize parallel computation across 10 cores for the doseRider model to enhance efficiency. Critical parameters include specifying the `dose_col`, which identifies the column containing the dose information essential for the analysis. Due to the wide range of dose values, a log10 transformation (`log_transform = TRUE`) is applied to reduce variability and stabilize the model for a more consistent dose-response relationship. The gene sets are filtered based on size, retaining only those between 20 and 500 genes, and the analysis is conducted using a linear mixed model (LMM). The false discovery rate (FDR) method is used for multiple testing correction, ensuring robust pathway selection through pathway filtering. This approach is applied to RNA-Seq data without additional covariates.

```{r}
# Default filtering parameters
default_params <- list(
    pca = list(pca_threshold = 0.7),
    variance = list(variance_threshold = 0.1),
    correlation = list(correlation_threshold = 0.4)
  )

spline_knots = 0

# Run doseRider analysis
dose_rider_results <- DoseRiderParallel(
  se = se, 
  gmt = gmt, 
  dose_col = "Dose", 
  omic = "rnaseq", 
  minGSsize = 50, 
  maxGSsize = 200, 
  method = "fdr", 
  covariates = c(),
  modelType = "LMM", 
  num_cores = 20,
  filters = c("variance","pca","correlation"),
  filter_params = default_params, # If empty took the default anyway
  log_transform = T,
  models = c("linear","non_linear_mixed"),
  spline_knots = spline_knots,
  knot_method = "quantile",
  
)


```


In this example, the `doseRider` model allows filtering of results based on specified criteria. Here, we use the `filter_DoseRider()` function to filter the results based on the false discovery rate (FDR) with a threshold of 0.1. The `model_type` is set to `"non_linear_mixed"`, ensuring that only significant pathways or gene sets identified by the non-linear mixed model are retained. This filtering step is crucial for refining the output and focusing on the most statistically relevant results in the context of dose-response analysis.

```{r}
# Filter doseRider results based on FDR
dose_rider_results_filter <- filter_DoseRider(
  dose_rider_results, 
  model_type = "non_linear", 
  filter_type = "fdr", 
  threshold = 0.1
)

# Convert results to data frame
res_df <- as.data.frame.DoseRider(dose_rider_results)
table(res_df$best_model)
```

### 5. Compute BMD interval confident

In this example, `doseRider` internally computes the benchmark dose (BMD) for significant pathways as part of its standard analysis. However, an additional function, `compute_bmd_bounds_parallel()`, was developed to estimate the confidence intervals for the BMD using bootstrapping. This process enhances the reliability of the BMD estimates by accounting for variability in the data. To reduce computational demand, we perform only 10 bootstrapping iterations (`n_bootstrap = 25`) in this case, though typically a larger number would be used for more robust confidence intervals. The computation is parallelized across 100 cores to speed up the process.

```{r}
#Compute BMD bounds
bmd_bounds_df <- compute_bmd_bounds_parallel(
  dose_rider_results = dose_rider_results_filter,
  dose_col = "log_Dose",
  sample_col = "sample",
  covariates = c(),
  omic = "rnaseq",
  n_bootstrap = 25,
  num_cores = 10,
  clusterResults = T
)

#knitr::kable(bmd_bounds_df, digits = 3)
```

### 6. Results and Discussion

Generate and save plots based on the DoseRider analysis.

```{r}
# Set plotting parameters and output path
top <- 10
save_path <- "~/doseRider/plots/"
dir.create(save_path, showWarnings = FALSE)
PLOT_WIDTH <- 11
PLOT_HEIGHT <- 11
units <- "cm"
```



#### 6.1 Dataset-Level Visualizations

```{r}
# 1. Dotplot of Top Pathways
p1 <- plot_dotplot_top_pathways(dose_rider_results_filter, top = top)
ggsave(paste0(save_path, "plot1_dotplot_top_pathways.jpeg"), plot = p1, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)

# 2. Dose Response Heatmap
p2 <- dose_response_heatmap(dose_rider_results_filter, dose_col = "Dose", top = top)
jpeg(file = paste0(save_path, "plot2_dose_response_heatmap.jpeg"), width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, res = 600)
plot(p2, heatmap_legend_side = "bottom", annotation_legend_side = "bottom", padding = grid::unit(c(1, 1, 1, 1), "cm"))
dev.off()

# 3. Gene Set Random Effects
p3 <- plot_gene_set_random_effects(dose_rider_results_filter, dose_col = "log_Dose", top = top)
ggsave(paste0(save_path, "plot3_gene_set_random_effects.jpeg"), plot = p3, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)

# 4. Top Pathway Responses
p4 <- plot_top_pathway_responses(dose_rider_results_filter, top = 8, ncol = 2, text_size = 5, dose_col = "log_Dose", clusterResults = TRUE, scale_values = FALSE, center_values = TRUE)
ggsave(paste0(save_path, "plot4_top_pathway_responses.jpeg"), plot = p4, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)

# 5. BMD Density and Peaks
data_bmd <- get_bmd_range(dose_rider_results = dose_rider_results_filter)
p5 <- plot_bmd_density_and_peaks(data_bmd)
ggsave(paste0(save_path, "plot5_bmd_density_and_peaks.jpeg"), plot = p5, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)

# 6. TCD Density and Peaks
data_tcd <- get_tcd_range(dose_rider_results = dose_rider_results_filter)
p6 <- plot_tcd_density(data_tcd)
ggsave(paste0(save_path, "plot6_tcd_density_and_peaks.jpeg"), plot = p6, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)

# 7. BMD Confidence Intervals
p7 <- plot_bmd_confidence_intervals(head(bmd_bounds_df, 20))
ggsave(paste0(save_path, "plot7_bmd_confidence_intervals.jpeg"), plot = p7, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)

# 8. TCD Confidence Intervals
p8 <- plot_tcd1_confidence_intervals(bmd_bounds_df)
ggsave(paste0(save_path, "plot8_tcd_confidence_intervals.jpeg"), plot = p8, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)
```

---

#### Pathway-Level Visualizations

```{r} 
# 9. Gene Random Effect Relationship (example pathway)
p9 <- plot_gene_random_effect_relationship(dose_rider_results, "PI3K-Akt signaling pathway - Homo sapiens (human)")
ggsave(paste0(save_path, "plot9_gene_random_effect_relationship.jpeg"), plot = p9, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, dpi = 600)

# 10. Gene-Level Heatmap (example pathway)
p10 <- create_gene_heatmap(dose_rider_results_filter, dose_col = "Dose", gene_set_name = "PI3K-Akt signaling pathway - Homo sapiens (human)")
jpeg(file = paste0(save_path, "plot10_gene_heatmap.jpeg"), width = PLOT_WIDTH, height = PLOT_HEIGHT, units = units, res = 600)
plot(p10, heatmap_legend_side = "bottom", annotation_legend_side = "bottom", padding = grid::unit(c(1, 1, 1, 1), "cm"))
dev.off()
```





