---
title: "Using doseRider for studying Non-Linear Dose-Response in BPA."
author: "Pablo Monfort-Lanzas"
date: "2023-08-01"
output: html_document
---

# Using doseRider for studying Non-Linear Dose-Response in BPA

The doseRider package is a powerful tool for gene expression data analysis, making use of Generalized Additive Mixed Models (GAMMs) to study the relationship between dose-response and gene sets. This analysis will employ the doseRider function to evaluate transcriptomic data from a study on bisphenol A (BPA) and 15 BPA alternative chemicals on cultured breast cancer cells (MCF-7).

We have preprocessed the data from study PRJNA869442, available on NCBI's Sequence Read Archive (SRA), locally. The goal is to use doseRider to detect potential non-linear dose-response relationships within the dataset.

## RMarkdown Workflow for doseRider Analysis

1. **Loading required packages**

```{r, warning=FALSE, echo=FALSE, include=FALSE}
library(doseRider)
library(DESeq2)
library(ggplot2)
```

2. **Data import**


The dataset PRJNA869442 (GSE211183) was provided by a toxicogenomics study that aimed to evaluate potential hazards and compare potencies of bisphenol A (BPA) and 15 BPA alternative chemicals in cultured breast cancer cells. The study utilized MCF-7 cells and exposed them to BPA and the alternative chemicals at different concentrations ranging from 0.0005 µM to 100 µM for 48 hours. The dataset includes a total of 768 samples, which include technical controls, solvent controls (0.1% DMSO), and 10 concentrations of each bisphenol chemical tested in quadriplicate. In addition to the test chemicals, two reference chemicals, estradiol and dexamethasone, were also included in the study. The transcriptomic profiling of gene expression data was performed using TempO-Seq, a high-throughput assay that allows for targeted transcriptomic analysis of up to 6,144 samples in one run, without the need for RNA purification, cDNA synthesis, or capture of targeted RNA. TempO-Seq exhibits 99.6% specificity, single-cell sensitivity, and excellent correlation with fold differences measured by RNA-Seq for 20,629 target genes, making it a powerful tool for toxicological risk assessment and hazard evaluation of data-poor chemicals.

```{r}
load("../data/PRJNA869442.rda")
```

For each compound evaluated in the study, we carefully selected the doses based on multiple factors, including the presence of visible precipitation, activating two or more stress response biomarkers, and significant decreases in cell proliferation. These criteria were informed by additional results obtained from the publication and other relevant data. By considering these various factors, we ensured that the chosen doses were appropriate for the analysis and effectively captured the potential hazards and toxicological responses of the compounds under investigation. This meticulous dose selection process enhances the reliability and validity of our study's findings and allows for a comprehensive comparison of the compounds' potencies and gene expression responses.

```{r}
# Create a list to represent the compound doses
compound_doses <- list(
  "2,4'-BPF" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "2,4'-BPS" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10, 50, 100),
  "4,4'-BPF" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10, 50),
  "BADGE" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "BPA" = c(0.0005, 0.1, 0.5, 1, 5, 10),
  "BPAF" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "BPAP" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5),
  "BPC" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "BPS" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10, 50, 100),
  "BPS-MAE" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "BPS-MPE" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "BTUM" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "D-8" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10),
  "DDS" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5),
  "P201" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10, 50),
  "TGSA" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5, 10)
)

```

**Filtering doses for each compound**

```{r}
# Convert the 'dose:ch1' column of the PRJNA869442 dataframe to numeric and store the result in a new column named 'dose'
PRJNA869442$dose <- unlist(as.numeric(PRJNA869442$`dose:ch1`))

# Filter the PRJNA869442 dataframe to retain only those rows where the 'title' column contains the string "BPAF"
bps <- PRJNA869442[, grepl("BPS", PRJNA869442$title)]
# Further filter the 'bpaf' dataframe to include only rows where the 'dose' matches values in 'compound_doses$BPAF'
bps <- bps[, bps$dose %in% compound_doses$BPS]

# Prepare the data for differential expression analysis:
# Create a DESeqDataSet object using the count data in 'bps' and its corresponding sample data with a specific design formula
dds <- DESeqDataSetFromMatrix(countData = assay(bps),
                              colData   = colData(bps),
                              design    = ~ dose + dose*dose)

# (The next line, if uncommented, would filter the 'dds' dataset to retain only rows where the total count across all samples is greater than 1)
# dds <- dds[rowSums(counts(dds)) > 1]

# Perform differential expression analysis on the 'dds' dataset using DESeq, and execute the analysis in parallel for efficiency
dds <- DESeq(dds, parallel = TRUE)
# Extract and store the results of the differential expression analysis in the 'res' object
res <- results(dds)

# Apply a variance-stabilizing transformation to the count data in 'dds' and store the result in 'counts_normal'
counts_normal <- assay(varianceStabilizingTransformation(dds))
# Replace the original count data in 'bpaf' with the variance-stabilized counts from 'counts_normal'
assay(bps) <- counts_normal
# Create a new 'sample' column in 'bpaf' that duplicates the values from its 'description' column
bps$sample <- bps$description
# Apply a log transformation to the 'dose' column in 'bpaf' for potential further analysis or visualization
#bps$dose <- log(bps$dose + 1)
# Filter 'bps' to retain only those genes with a baseMean value greater than 20
genes_to_keep <- rownames(res[res$baseMean > 20, ])
bps <- bps[genes_to_keep, ]

```

3. **Load the genesets**

```{r}

gmt <- loadCPDB("Symbol")
gmt <- filter_gmt_by_size(gmt = gmt, minGenesetSize = 15, maxGenesetSize = 1250)  
```

3. **Setting the dose-response model**

doseRider provides support for both Linear Mixed Models (LMM) and General Additive Mixed Models (GAMM). For this example, we will utilize `doseRiderLMM` due to its faster computation. While GAMM models offer a versatile approach, they are more computationally intensive, making them less ideal for modeling the entire consensus pathway database, especially with gene sets ranging from 15 to 1,250 genes (specified as `minGenesetSize = 15` and `maxGenesetSize = 1250`). Within the LMM framework in doseRider, three distinct models are calculated: a base model with only the gene as a random effect, a linear model with the gene as a random effect, and a model incorporating cubic splines.


```{r}
res <- DoseRiderLMM(bps, gmt = gmt, covariate = "sample", omic = "proteomics", minGSsize = 15, maxGSsize = 1250)
resDf <- as.data.frame.DoseRider(res)
table(unlist(resDf["Adjusted_Cubic_P_Value"]) < 0.05)
resDf <- resDf[unlist(resDf["P_Value_Cubic"]) < 0.05 & unlist(resDf["P_Value_Linear"] < 0.05),]
```

5. **Plot significant geneset**

```{r}
library(cowplot)

# Create a list to store the plots
plot_list <- list()

for (geneset_sign in resDf[resDf$Adjusted_Cubic_P_Value < 0.05,]$Geneset) {
  p <- plot_smooth(res, geneset_sign, center_values = TRUE) + xlab("log(dose)")
  plot_list[[geneset_sign]] <- p
}

# Combine the plots into a single grid plot
plot_grid(plotlist = plot_list, ncol = 3)

```

6. **Visualizing results**

doseRider provides diverse plot options to understand data and model fits better.

```R
plot(results)
```

7. **Saving the results**

```R
write.csv(results, file = "doseRider_results.csv")
```

By following this workflow, you can investigate non-linear dose-response relationships in the high-throughput transcriptomic data. This could lead to vital insights about the impacts of BPA and its alternative chemicals on cellular processes, thereby informing risk assessment of these understudied chemicals.

**Note:** The specifics of this RMarkdown document may need adjustments based on the actual preprocessed data. Always ensure a thorough data validation before proceeding with the analysis.
