---
title: "Using doseRider for studying Non-Linear Dose-Response in BPA."
author: "Pablo Monfort-Lanzas"
date: "2023-08-01"
output: html_document
---

# Using doseRider for studying Non-Linear Dose-Response in BPA

The `doseRider` package is used for analyzing dose-response relationships in gene expression data using Generalized Additive Mixed Models (GAMMs). This report focuses on bisphenol A (BPA) and its alternatives, analyzing their impact on gene expression in MCF-7 cells.

## RMarkdown Workflow for doseRider Analysis

### 1. Loading required packages

```{r, warning=FALSE, echo=FALSE, include=FALSE}
library(doseRider)
library(DESeq2)
library(ggplot2)
library(dplyr)
library(edgeR)
```

### 2. Data Import

Data from PRJNA869442 (GSE211183) has been preprocessed for this analysis.

```{r}
load("../data/PRJNA869442.rda")
```

#### Dose Selection for Each Compound

Doses were selected based on criteria like precipitation, stress response biomarkers, and cell proliferation.

```{r}
compound_doses <- list(
  "BPAF" = c(0.0005, 0.001, 0.01, 0.1, 0.5, 1, 5)
)
  doses <- c(0,compound_doses["BPAF"][[1]])
  
# Example filtering for BPA
bpaf_data <- PRJNA869442[,(colData(PRJNA869442)$Chemical %in% c("Cells, no treatment","BPAF")) &
                             (colData(PRJNA869442)$Dose %in% doses)]
bpaf_data$sample <- colnames(bpaf_data)
```

### 3. Differential Expression Analysis

```{r}
colData(bpaf_data)$Dose <- unlist(colData(bpaf_data)$Dose)
#colData(bpa_data)$Dose <- log(unlist(colData(bpa_data)$Dose) + 1)
dds <- DESeqDataSetFromMatrix(countData = assay(bpaf_data),
                              colData   = colData(bpaf_data),
                              design    = ~ Dose*Dose)

dds <- DESeq(dds, parallel = TRUE, quiet = TRUE)
res <- results(dds)

filter_res <- as.data.frame(res[(res$baseMean > 11) & (res$pvalue < 0.1),])
bpaf_data <- estimate_model_parameters(bpaf_data)

# Filter low express genes
filter_bpaf_data <- bpaf_data[rownames(bpaf_data) %in% rownames(filter_res),]

# Extract metadata from the filtered data
metadata <- colData(filter_bpaf_data)

# Extract expression data from the filtered data
expression_data <- assay(filter_bpaf_data)

# Save metadata to a TSV file
write.table(metadata, "metadata.tsv", sep="\t", quote=FALSE)

# Save expression data to a TSV file
write.table(expression_data, "expression_data.tsv", sep="\t", quote=FALSE)

```

### 4. Load Gene Sets

```{r}
# gmt <- loadCPDB("Symbol")
# top_geneset <- read.csv("../../Projects/TOX/ToxicoDataset/results/TG-GATES/WeightedScoreCPDBTGGATES.tsv", sep = ",")
# gmt_id <- doseRider:::filter_gmt_by_id(gmt, unique(top_geneset[abs(top_geneset$WeightedScore) > 1,]$X))

# Specify the path to your GMT file
gmt_h_path <- "../external/h.all.v2023.2.Hs.symbols.gmt" 
gmt_perturbation_path <- "../external/c2.cgp.v2023.2.Hs.symbols.gmt"

# Use the function to read the GMT file
gmt <- filter_gmt_by_size(read_gmt(gmt_perturbation_path),20,200)
#gmt_h_data <- filter_gmt_by_size(read_gmt(gmt_h_path),20,100)
#gmt_data <- c(gmt_h_data, gmt_p_data)
```

### 5. Dose-Response Model with doseRider
```{r}
#' Compute BMD Bounds using Bootstrapping with Parallel Gene Set Processing
#'
#' This function calculates the lower and upper bounds of the Benchmark Dose (BMD)
#' using bootstrapping on the raw data for significant gene sets, with the gene set
#' processing parallelized across multiple cores.
#'
#' @param dose_rider_results A DoseRider object containing results of the analysis.
#' @param dose_col Name of the column representing dose information.
#' @param sample_col Name of the column representing sample information.
#' @param covariates Optional, vector specifying the covariate column(s) in `se`.
#' @param omic Type of omics data, defaults to "rnaseq".
#' @param n_bootstrap The number of bootstrap samples to generate. Default is 1000.
#' @param ci_level Confidence interval level, default is 0.95.
#' @param num_cores The number of cores to use for parallel processing of gene sets. Default is 5.
#' @import foreach
#' @import doParallel
#' @import parallel
#' @import progress
#' @return A data frame containing the lower and upper BMD bounds for each significant gene set.
#'
#' @examples
#' \dontrun{
#' data("SummarizedExperiment")
#' gmt <- list(geneSet1 = list(genes = c("gene1", "gene2")))
#' results <- DoseRider(se, gmt, "dose", "sample", "covariate", "rnaseq", modelType = "GAMM")
#' bmd_bounds <- compute_bmd_bounds_parallel(results, "dose", "sample", n_bootstrap = 1000, ci_level = 0.95, num_cores = 4)
#' print(bmd_bounds)
#' }
#'
#' @export
compute_bmd_bounds_parallel <- function(dose_rider_results, dose_col = "dose", sample_col = "sample", ci_level = 0.95,
                                        covariates = c(), omic = "rnaseq", n_bootstrap = 1000, num_cores = 5) {

  # Register the parallel backend for gene set processing
  cl <- makeCluster(num_cores)
  registerDoParallel(cl)

  # Initialize progress bar
  pb <- txtProgressBar(min = 0, max = length(dose_rider_results), style = 3)

  # Perform parallel processing of gene sets
  bmd_bounds_list <- foreach(geneset_idx = seq_along(dose_rider_results), .combine = rbind, .packages = c("lme4", "doseRider", "dplyr")) %dopar% {
    geneset_name <- names(dose_rider_results)[geneset_idx]
    geneset <- dose_rider_results[[geneset_name]]
    long_df <- geneset$Raw_Values[[1]]
    best_model <- geneset$best_model
    formula <- doseRider:::create_lmm_formula("counts", dose_col, "gene", covariates, best_model, omic)
    bmd <- extract_bmd_for_pathway(dose_rider_results, geneset_name)

    # Perform bootstrap sampling
    bmd_values <- replicate(n_bootstrap, {
      bootstrap_indices <- sample.int(n = nrow(long_df), size = nrow(long_df)*0.8, replace = TRUE)
      long_df_bootstrap <- long_df[bootstrap_indices, , drop = FALSE]
      bootstrap_results <- suppressMessages(doseRider:::fit_model_compute_bmd(long_df_bootstrap, formula, omic = omic, clusterResults = FALSE, dose_col = dose_col))
      if (!is.na(bootstrap_results)) {
        return(unlist(bootstrap_results$AllGenes)[1])
      }
      return(NA)
    })

    # Remove NA values from BMD calculations
    bmd_values <- na.omit(bmd_values)

    if (length(bmd_values) > 0) {
      # Calculate the lower and upper bounds for the BMD
      lower_bound <- quantile(bmd_values, probs = (1 - ci_level) / 2)
      upper_bound <- quantile(bmd_values, probs = 1 - (1 - ci_level) / 2)
      mean_bmd <- mean(bmd_values)
      median_bmd <- median(bmd_values)

      # Create a data frame with the results
      result <- data.frame(
        Geneset = geneset_name,
        Lower_Bound = lower_bound,
        Upper_Bound = upper_bound,
        Mean_BMD = mean_bmd,
        Median_BMD = median_bmd,
        Best_Model = best_model,
        stringsAsFactors = FALSE
      )
    } else {
      result <- NULL
    }

    # Update progress bar
    setTxtProgressBar(pb, geneset_idx)

    return(result)
  }

  # Stop the parallel backend
  stopCluster(cl)
  close(pb)

  return(bmd_bounds_list)
}

```
```{r}
dose_rider_results <- DoseRiderParallel(se = filter_bpaf_data, gmt = gmt, 
                                           dose_col = "Dose", omic = "rnaseq", 
                                           minGSsize = 20, maxGSsize = 200, 
                                           method = "bonferroni", covariates = c(),
                                          modelType = "LMM", num_cores = 4,
                                          clusterResults = T, FilterPathway = F)

res_df <- as.data.frame.DoseRider(dose_rider_results)
table(res_df$best_model)


#save(res, file = "../../doseRider_save/res_bpa_20_1000.rda")
dose_rider_results_filter <- head(filter_DoseRider(dose_rider_results, model_type = "all", filter_type = "fdr", threshold = 0.001), 100)
# Measure time for the updated function (parallel gene set processing)
time_updated <- system.time({
  bmd_bounds_df <- doseRider::compute_bmd_bounds_parallel(
    dose_rider_results = dose_rider_results_filter, 
    dose_col = "Dose", 
    sample_col = "sample", 
    covariates = c(), 
    omic = "rnaseq", 
    n_bootstrap = 25, 
    num_cores = 6
  )
})
print(time_updated)
```

```{r}
# Measure time for the updated function (parallel gene set processing)
time_updated <- system.time({
  bmd_bounds_df <- compute_bmd_bounds_parallel(
    dose_rider_results = dose_rider_results_filter, 
    dose_col = "Dose", 
    sample_col = "sample", 
    covariates = c(), 
    omic = "rnaseq", 
    n_bootstrap = 25, 
    num_cores = 6
  )
})
print(time_updated)

```
## Results and Discussion

```{r}
top <- 5
save_path <- "/home/monfortl/doseRider/plots/"
#dir.create(save_path)
PLOT_HEIGHT <- 2500
PLOT_WIDTH <- 2500

plot_width <- PLOT_WIDTH
plot_height <- PLOT_WIDTH


p1 <- dose_response_heatmap(dose_rider_results, dose_col = "Dose", top = top, order_column = "best_model_pvalue", decreasing = F )
plotFile1 <- paste0(save_path,"/plot1.jpeg")

p2 <- plot_gene_set_random_effects(dose_rider_results, dose_col = "Dose",order_column = "best_model_pvalue", top = top)
plotFile2 <- paste0(save_path,"/plot2.jpeg")

p3 <- plot_top_pathway_responses(dose_rider_results, top = 6, ncol = 2, order_column = "best_model_pvalue", text_size = 5)
plotFile3 <- paste0(save_path,"/plot3.jpeg")

p4 <- plot_gene_random_effect_relationship(dose_rider_results, "BENPORATH_ES_CORE_NINE_CORRELATED")
plotFile4 <- paste0(save_path,"/plot4.jpeg")

p5 <- plot_dotplot_top_pathways(dose_rider_results, top = top, order_column = "Genes", decreasing = T)
plotFile5 <- paste0(save_path,"/plot5.jpeg")

p6 <- create_gene_heatmap(dose_rider_results,dose_col = "Dose", gene_set_name = "MENSE_HYPOXIA_UP")
plotFile6 <- paste0(save_path,"/plot6.jpeg")

data_bmd <- get_bmd_range(dose_rider_results = dose_rider_results)
p7 <- plot_bmd_density_and_peaks(data_bmd)
plotFile7 <- paste0(save_path,"/plot7.jpeg")

p10 <- plot_bmd_confidence_intervals(head(bmd_bounds_df,20))
plotFile10 <- paste0(save_path,"/plot10.jpeg")


data_tcd <- get_tcd_range(dose_rider_results = filter_DoseRider(dose_rider_results, model_type = "non_linear"))
p8 <- plot_tcd_density(data_tcd)
plotFile8 <- paste0(save_path,"/plot8.jpeg")

jpeg(file=plotFile1, width = PLOT_WIDTH - 1500, height = PLOT_HEIGHT - 1500, units = "px")
  plot(p1, heatmap_legend_side = "bottom", annotation_legend_side = "bottom",padding = unit(c(1, 1, 1, 1), "cm"))
dev.off()


ggsave(plotFile2, plot = p2, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = "px", dpi = 600)

ggsave(plotFile3, plot = p3, width = plot_width, height = plot_height+1000, units = "px", dpi = 600)

ggsave(plotFile4, plot = p4, width = PLOT_WIDTH+1800, height = PLOT_HEIGHT+600, units = "px", dpi = 600)

ggsave(plotFile5, plot = p5, width = PLOT_WIDTH + 500, height = PLOT_HEIGHT + 500, units = "px", dpi = 600)

jpeg(file=plotFile6, width = PLOT_WIDTH - 1500, height = PLOT_HEIGHT - 1500, units = "px")
 plot(p6, heatmap_legend_side = "bottom", annotation_legend_side = "bottom",padding = unit(c(1, 1, 1, 1), "cm"))
dev.off()

ggsave(plotFile7, plot = p7, width = PLOT_WIDTH, height = PLOT_HEIGHT, units = "px", dpi = 600)

ggsave(plotFile10, plot = p10, width = PLOT_WIDTH, height = PLOT_HEIGHT+500, units = "px", dpi = 600)

```

# Microarray example:

dataset1: Human Transcriptomic Integrated Approach for Testing and Assessment Oral Reference Dose for Steatosis: 0.00, 0.3125, 0.625, 1.25, 2.50, 5.00 (24 samples) GSE135117

dataset2: Primary hepatocyte cultures from 3 human donors were treated with TCDD and PCB 126 compounds, with 14 samples per compound (6 doses) (6.5 to 14 log10 M doses for TCDD and 5 to 12 log10 M doses for PCB 126). GSE14553

dataset3: The cell used in this study is THP-1 macrophages from Homo sapiens. They were exposed to rigid multi-walled carbon nanotubes (rCNT) at three concentrations (5, 10, or 20 Âµg/mL) and untreated controls at 24, 48, and 72 hours, with each group performed as triplicates. GSE146708


The cell used is BEAS-2B bronchial epithelial cells (Homo sapiens), exposed to N-hydroxylated PhIP at nine different doses ranging from 1 x 10-6 to 5 x 10-10 M, with triplicate cell cultures for each dose. There is no mention of donors in this dataset. PRJNA151269; GEO: GSE34635

```{r}
lColor <- c(
  "#37475b", "#90a9c6", "#f0f8f9", "#a2a9bb", "#a6bbd0", "#686872", "#c98768", "#c98768",
  "#D32F2F", "#388E3C", "#1976D2", "#FFC107", "#7C4DFF", "#E64A19", "#009688", "#FBC02D"
)
write_gmt <- function(data_list, file_name) {
  gmt_lines <- sapply(data_list, function(entry) {
    pathway <- entry$pathway
    genes <- entry$genes
    if (is.null(genes)) {
      genes <- c("")
    } else {
      genes <- genes[!is.na(genes)]  # Remove any NA values
    }
    c(pathway, "", genes)
  }, simplify = FALSE)
  
  gmt_content <- sapply(gmt_lines, function(line) {
    paste(line, collapse = "\t")
  })
  
  file_path <- file.path("external_data", file_name)
  
  writeLines(gmt_content, con = file_path)
}

convert_and_write_gmt <- function(data_list, file_name) {
  if (!dir.exists("external_data")) {
    dir.create("external_data")

  
  write_gmt(data_list, file_name)
}

  
  # Load the datasets
  data("CPDB_pathways_genes_ensembl", package = "doseRider")
  data("CPDB_pathways_genes_entrez", package = "doseRider")
  data("CPDB_pathways_genes_refseq", package = "doseRider")
  data("CPDB_pathways_genes_symbol", package = "doseRider")
  data("CPDB_pathways_metabolites", package = "doseRider")
  
  # Write each dataset to GMT format
  write_gmt(CPDB_pathways_genes_ensembl, "CPDB_pathways_genes_ensembl.gmt")
  write_gmt(CPDB_pathways_genes_entrez, "CPDB_pathways_genes_entrez.gmt")
  write_gmt(CPDB_pathways_genes_refseq, "CPDB_pathways_genes_refseq.gmt")
  write_gmt(CPDB_pathways_genes_symbol, "CPDB_pathways_genes_symbol.gmt")
  write_gmt(CPDB_pathways_metabolites, "CPDB_pathways_metabolites.gmt")
}

convert_and_write_gmt()


```
