---
title: "Using doseRider for studying Non-Linear Dose-Response in BPA."
author: "Pablo Monfort-Lanzas"
date: "2023-08-01"
output: html_document
---

# Using doseRider for studying Non-Linear Dose-Response in BPA

The `doseRider` package is used for analyzing dose-response relationships in gene expression data using Generalized Additive Mixed Models (GAMMs). This report focuses on bisphenol A (BPA) and its alternatives, analyzing their impact on gene expression in MCF-7 cells.

## RMarkdown Workflow for doseRider Analysis

### 1. Loading required packages

```{r, warning=FALSE, echo=FALSE, include=FALSE}
library(doseRider)
library(DESeq2)
library(ggplot2)
library(dplyr)
library(SummarizedExperiment)
```

### 2. Data Import

Data from PRJNA869442 (GSE211183) has been preprocessed for this analysis.

```{r}
load("../data/PRJNA869442.rda")
```

#### Dose Selection for Each Compound

Doses were selected based on criteria like precipitation, stress response biomarkers, and cell proliferation.

```{r}
compound_doses <- list(
  "BPA" = c(0.0005, 0.1, 0.5, 1)
)
  doses <- c(0,compound_doses["BPA"][[1]])
  
# Example filtering for BPA
bpa_data <- PRJNA869442[,(colData(PRJNA869442)$Chemical %in% c("Cells, no treatment","BPA")) &
                             (colData(PRJNA869442)$Dose %in% doses)]
bpa_data$sample <- colnames(bpa_data)
```

### 3. Differential Expression Analysis

```{r}
colData(bpa_data)$Dose <- unlist(colData(bpa_data)$Dose)
#colData(bpa_data)$Dose <- log(unlist(colData(bpa_data)$Dose) + 1)
dds <- DESeqDataSetFromMatrix(countData = assay(bpa_data),
                              colData   = colData(bpa_data),
                              design    = ~ Dose*Dose)

dds <- DESeq(dds, parallel = TRUE, quiet = TRUE)
res <- results(dds)

filter_res <- as.data.frame(res[(res$baseMean > 11) & (res$padj < 0.1),])
bpa_data <- estimate_model_parameters(bpa_data)

# Filter low express genes
filter_bpa_data <- bpa_data[rownames(bpa_data) %in% rownames(filter_res),]
```

### 4. Load Gene Sets

```{r}
# Function to read a GMT file and convert it into a structured list
read_gmt <- function(file_path) {
  # Open the file for reading
  con <- file(file_path, "r")
  
  # Initialize an empty list to store the pathways and genes
  gmt <- list()
  
  # Read the file line by line
  while(TRUE) {
    line <- readLines(con, n = 1, warn = FALSE)
    if(length(line) == 0) { # Check if end of file
      break
    }
    
    # Split the line into components
    elements <- strsplit(line, "\t")[[1]]
    
    # The first element is the pathway name, the rest are genes
    pathway_name <- elements[1]
    genes <- elements[-c(1, 2)] # Exclude pathway name and URL
    
    # Add to the list
    gmt[[length(gmt) + 1]] <- list(pathway = pathway_name, genes = genes)
  }
  
  # Close the file connection
  close(con)
  
  return(gmt)
}

# Specify the path to your GMT file
gmt_h_path <- "../external/h.all.v2023.2.Hs.symbols.gmt" 
gmt_perturbation_path <- "../external/c2.cgp.v2023.2.Hs.symbols.gmt"

# Use the function to read the GMT file
gmt_p_data <- read_gmt(gmt_perturbation_path)
gmt_h_data <- read_gmt(gmt_h_path)
gmt_data <- c(gmt_h_data, gmt_p_data)
# Optionally, print the first few pathways to check
if (length(gmt_data) > 0) {
  print(gmt_data[[1]])
  print(gmt_data[[2]])
}

```

## Try several formula to check which represent better the data

### 5. Dose-Response Model with doseRider

```{r}
se <- filter_bpa_data
dose_col <- "Dose"
sample_col <- "sample"
omic <- "rnaseq"
for (i in c(1:length(gmt_data))) {

    
  geneset <- gmt_data[[i]]$genes
  geneset_pathway <- gmt_data[[i]]$pathway
  cat("\n[+]",geneset_pathway)
  long_df <- suppressWarnings(doseRider:::prepare_data(se=se, geneset=geneset, dose_col=dose_col, sample_col=sample_col, omic=omic))
    # Check gene set size criteria
  if (!is.null(long_df) & length(unique(long_df$gene)) > 5 & length(unique(long_df$gene)) < 1000) {

  non_linear_formula_tcgsa <- "counts ~ 1 + (1|gene) + bs(Dose) + (0 + bs(Dose)|gene) + offset(log(size_factor))"
  non_linear_formula_cluster <-  "counts ~ bs(Dose) + (bs(Dose)|gene) + offset(log(size_factor))"
  non_linear_formula_doserider <- "counts ~ (bs(Dose) | gene) + offset(log(size_factor))"
  non_linear_formula_timeomics <- "counts ~ (bs(Dose)) + (Dose|gene) + offset(log(size_factor))"
  non_linear_formula_old_DoseRider <-  "counts ~ bs(Dose) + (1 + Dose | gene) + offset(log(size_factor))"

  
  # Fit the model
  non_linear_results_tcgsa <- suppressWarnings(doseRider:::fit_lmm(non_linear_formula_tcgsa, long_df, omic))
  non_linear_results_cluster <- suppressWarnings(doseRider:::fit_lmm(non_linear_formula_cluster, long_df, omic))
  non_linear_results_doserider <- suppressWarnings(doseRider:::fit_lmm(non_linear_formula_doserider, long_df, omic))
  non_linear_results_timeomics <- suppressWarnings(doseRider:::fit_lmm(non_linear_formula_timeomics, long_df, omic))
  non_linear_results_timeomics_old_DoseRider <- suppressWarnings(doseRider:::fit_lmm(non_linear_formula_old_DoseRider, long_df, omic))
  
  # Normalize
  smooth_values <- doseRider:::smooth_pathway_trend(non_linear_results_tcgsa, long_df, dose_col, sample_col, omic, TRUE, c(), dose_points = 50)
  smooth_values <- aggregate(predictions ~ Dose + gene, data = smooth_values, FUN = mean)
  colnames(smooth_values) <- c("Dose", "gene","TCGSA")

  # Aggregate Cluster predictions
  cluster_predictions <- doseRider:::smooth_pathway_trend(non_linear_results_cluster, long_df, dose_col, sample_col, omic, TRUE, c(), dose_points = 50)
  cluster_mean <- aggregate(predictions ~ Dose + gene, data = cluster_predictions, FUN = mean)
  smooth_values$Cluster <- cluster_mean$predictions
  
  # Aggregate DoseRider predictions
  doserider_predictions <- doseRider:::smooth_pathway_trend(non_linear_results_doserider, long_df, dose_col, sample_col, omic, TRUE, c(), dose_points = 50)
  doserider_mean <- aggregate(predictions ~ Dose + gene, data = doserider_predictions, FUN = mean)
  smooth_values$DoseRider <- doserider_mean$predictions
  
  # Aggregate TIMEOMICS predictions
  timeomics_predictions <- doseRider:::smooth_pathway_trend(non_linear_results_timeomics, long_df, dose_col, sample_col, omic, TRUE, c(), dose_points = 50)
  timeomics_mean <- aggregate(predictions ~ Dose + gene, data = timeomics_predictions, FUN = mean)
  smooth_values$TIMEOMICS <- timeomics_mean$predictions
  
  # Aggregate SMOOTH predictions
  smooth_predictions <- doseRider:::smooth_pathway_trend(non_linear_results_timeomics_old_DoseRider, long_df, dose_col, sample_col, omic, TRUE, c(), dose_points = 50)
  smooth_mean <- aggregate(predictions ~ Dose + gene, data = smooth_predictions, FUN = mean)
  smooth_values$OLD_DOSERIDER <- smooth_mean$predictions
  
  smooth_values <- smooth_values %>%
    group_by(gene) %>%
    mutate(
      TCGSA_centered = TCGSA - mean(TCGSA, na.rm = TRUE),
      Cluster_centered = Cluster - mean(Cluster, na.rm = TRUE),
      DoseRider_centered = DoseRider - mean(DoseRider, na.rm = TRUE),
      TIMEOMICS_centered = TIMEOMICS - mean(TIMEOMICS, na.rm = TRUE),
      OLD_DOSERIDER_centered = OLD_DOSERIDER - mean(OLD_DOSERIDER, na.rm = TRUE)
    ) %>%
    ungroup()
  
    long_df <- aggregate(counts ~ Dose + gene, data = long_df, FUN = mean)

    long_df <- long_df %>%
    group_by(gene) %>%
    mutate(
      counts = counts - mean(counts, na.rm = TRUE),

    ) %>%
    ungroup()
    
    library(ggplot2)

  # Function to create prediction plots
  plot_predictions <- function(data, dose_col, prediction_col, prediction_title) {
    return(ggplot(data, aes_string(x = dose_col, y = prediction_col, group = "gene", color = "'blue'")) +
      geom_line(size = 0.3) +
      labs(title = prediction_title, x = "Dose", y = "Predicted Expression") +
      theme_minimal() +
      theme(legend.position = 'none') +
      scale_color_manual(values = c("blue" = "blue")) # Ensures the color "blue" is used
    )
  }
  
  # Example usage
  p_raw <- plot_predictions(long_df, "Dose", "counts", "Real Counts Predictions")
  p_TCGSA <- plot_predictions(smooth_values, "Dose", "TCGSA_centered", "TCGSA Predictions")
  p_Cluster <- plot_predictions(smooth_values, "Dose", "Cluster_centered", "Cluster Predictions")
  p_DoseRider <- plot_predictions(smooth_values, "Dose", "DoseRider_centered", "DoseRider Predictions")
  p_TIMEOMICS <- plot_predictions(smooth_values, "Dose", "TIMEOMICS_centered", "TIMEOMICS Predictions")
  p_SMOOTH <- plot_predictions(smooth_values, "Dose", "OLD_DOSERIDER_centered", "OLD DOSERIDER Predictions")
  
  # Now, you can plot each of them or save them as needed

  
  # Display the plots
  p <- gridExtra::grid.arrange(p_raw, p_TCGSA, p_Cluster, p_DoseRider, p_TIMEOMICS,p_SMOOTH, ncol = 3)
  ggsave(p, file = paste0("../../doseRider_save/plots_compare/",geneset_pathway,".png"), width = 8, height = 5)
  }
  }
```
